ext/drreg/drreg.c

```diff
@@ -1831,115 +1831,198 @@ drreg_is_instr_spill_or_restore(void *drcontext, instr_t *instr, bool *spill OUT
-        pc = decode(drcontext, pc, &inst);
-
-        if (is_our_spill_or_restore(drcontext, &inst, &spill, &reg, &slot, &offs)) {
-            LOG(drcontext, DR_LOG_ALL, 3,
-                "%s @" PFX " found %s to %s offs=0x%x => slot %d\n", __FUNCTION__,
-                prev_pc, spill ? "spill" : "restore", get_register_name(reg), offs, slot);
+    /* XXX: if ilist not available, fall back to some best-effort logic. */
+    ASSERT(info->fragment_info.ilist != NULL, "ilist required for state restoration");
+    for (inst = instrlist_first(info->fragment_info.ilist);
+         inst && pc < info->raw_mcontext->pc; inst = instr_get_next(inst)) {
+        int len = instr_length(drcontext, inst);
+        pc += len;
+        ASSERT(pc <= info->raw_mcontext->pc, "went beyond fault pc while walking ilist");
+        bool app_gpr_restored_now = false;
+        bool app_aflags_restored_now = false;
+        bool app_aflags_written_to_reg_now = false;
+        /* Update book-keeping for gpr/aflags spill slot or aflags spill reg. */
+        if (is_our_spill_or_restore(drcontext, inst, &spill, &reg, &slot, &offs)) {
             if (spill) {
-                if (slot == AFLAGS_SLOT) {
-                    spilled_to_aflags = slot;
-                } else if (spilled_to[GPR_IDX(reg)] < MAX_SPILLS &&
-                           /* allow redundant spill */
-                           spilled_to[GPR_IDX(reg)] != slot) {
-                    /* This reg is already spilled: we assume that this new spill
-                     * is to a tmp slot for preserving the tool's value.
-                     */
+                /* Slot used in current instr may have a previous spilled gpr/aflags
+                 * value that may already have been restored. Forget that.
+                 */
+                forget_slot_use(gpr_spill_slot, &aflags_spill_slot, slot);
+                if (aflags_spill_reg == reg) {
+                    ASSERT(!gpr_native[GPR_IDX(aflags_spill_reg)],
+                           "reg with aflags cannot be native");
+                    aflags_spill_slot = slot;
+                } else if (gpr_native[GPR_IDX(reg)]) {
+                    gpr_spill_slot[GPR_IDX(reg)] = slot;
+                } else {
                     LOG(drcontext, DR_LOG_ALL, 3, "%s @" PFX ": ignoring tool spill\n",
                         __FUNCTION__, pc);
-                } else {
-                    spilled_to[GPR_IDX(reg)] = slot;
                 }
             } else {
-                if (slot == AFLAGS_SLOT && spilled_to_aflags == slot)
-                    spilled_to_aflags = MAX_SPILLS;
-                else if (spilled_to[GPR_IDX(reg)] == slot)
-                    spilled_to[GPR_IDX(reg)] = MAX_SPILLS;
-                else {
-                    LOG(drcontext, DR_LOG_ALL, 3, "%s @" PFX ": ignoring restore\n",
+                if (aflags_spill_slot == slot) {
+                    ASSERT(
+                        !gpr_native[GPR_IDX(reg)] ||
```

Comment:
I guess this may happen if the reg is dead and wasn't spilled before. I'll probably remove this assert.